proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,349 @@
+
+  module Normal where
+  import X
+  instance Show X where show _ = "normal"
+
+  module Splice where
+  import X
+  instance Show X where show _ = "splice"
+
+  module Bottom where
+  import X (X(..))
+  import splice X (X(..))
+  import Normal ()
+  import splice Splice ()
+  import splice Language.Haskell.TH.Lib ( stringE )
+
+  s1 = show MkX
+  s2 = $( stringE (show MkX) )
+
+This program is also rejected because the instances defined in ``Normal`` and ``Splice`` overlap.
+
+
+Other Considerations
+~~~~~~~~~~~~~~~~~~~~
+
+When ``TemplateHaskell`` is disabled, then ``ExplicitSpliceImports`` is a no-op.
+
+When ``TemplateHaskell`` is enabled but NOT ``ExplicitSpliceImports``, then all imports
+are implicitly additionally imported as splice imports, which matches the current behaviour.
+
+If the ``Prelude`` module is implicitly imported then it is also imported as a splice module. Hence the following is
+allowed::
+
+  zero = $(id [| 0 |])
+
+If ``NoImplicitPrelude`` is enabled then you have to import ``Prelude`` as a splice
+module as well in order to use names from ``Prelude`` in negative level splices::
+
+  {-# LANGUAGE TemplateHaskell #-}
+  {-# LANGUAGE ExplicitSpliceImports #-}
+  {-# LANGUAGE NoImplicitPrelude #-}
+
+  import splice Prelude
+
+  -- accepted
+  foo = $(id [|"foo"|])
+
+  -- rejected
+  foo = id $([|"foo"|])
+
+All exported names are at level 0. Splice imports can't be rexported, unless
+they are also imported normally.
+Allowing splice imports to be exported would turn a build-time only import into a runtime
+export. Maintaining the distinction between things only needed at build-time and
+things only needed at runtime allows project dependencies to be separated in the
+same way. This is important for cross-compilation.
+
+
+
+Drawbacks
+---------
+
+* The user has to be aware of the significance of using splice imports.
+
+
+
+Alternatives
+------------
+
+* ``splice`` imports could also bring identifiers into scope so that they
+  can be used everywhere in a module, not **only** in top-level splices as
+  the proposal suggest. This approach is not taken because it means that
+  build-time only dependencies can't be distinguished from runtime dependencies
+
+* Using a pragma rather than a syntactic modifier would fit in better with
+  how ``SOURCE`` imports work and make writing backwards compatible code easier::
+
+    import {-# SPLICE #-} B
+
+* It might be proposed that an alternative would be to work out which modules
+  need to be compiled based on usage inside a module. This would compromise the
+  principle that we can learn about what's needed for a module just by looking
+  at the import list in the module header.
+
+* The extension could only apply to **home** modules, because the benefits of
+  splice imports are when using GHC's ``--make`` mode. As the proposal stands,
+  for uniformity, any module used inside a top-level splice must be marked as
+  a splice module, even if it's an external module.
```

Comment:
You haven't touched on it in the motivation, but it's worth mentioning that restricting the distinction to home modules could be rather costly for monorepos with several package, as per https://github.com/ghc-proposals/ghc-proposals/pull/243#issuecomment-534125177 .